zk/zookeeper.go

```diff
@@ -0,0 +1,313 @@
+	PermRead   = zk.PermRead
+	PermWrite  = zk.PermWrite
+	PermCreate = zk.PermCreate
+	PermDelete = zk.PermDelete
+	PermAdmin  = zk.PermAdmin
+	PermAll    = zk.PermAll
+)
+
+var (
+	// ErrZnodeDoesNotExist is returned if the requested znode does not exist.
+	ErrZnodeDoesNotExist = errors.New("znode does not exist")
+)
+
+// ACL defines permissions, a scheme and an ID.
+type ACL struct {
+	Perms  int32
+	Scheme string
+	ID     string
+}
+
+// Config defines the default parameters for zookeeper setup.
+type Config struct {
+	// Address for zookeeper, an empty string will use DefaultAddr
+	Addr string
+
+	// If specified, all paths for the Client will be prefixed by this
+	BasePath string
+
+	// Authentication info for the connection
+	Auth *SchemeAuth
+
+	// Default ID to be used for Client.DefaultACL function
+	DefaultID *SchemeID
+}
+
+// SchemeAuth composes scheme and auth
+type SchemeAuth struct {
+	Scheme string
+	Auth   string
+}
+
+// ParseSchemeAuth parses input such as "scheme:auth"
+func ParseSchemeAuth(schemeAuth string) (*SchemeAuth, error) {
+	splits := strings.Split(schemeAuth, ":")
+	if len(splits) != 2 {
+		return nil, fmt.Errorf("schemeAuth expected format is 'schema:auth'")
+	}
+
+	return &SchemeAuth{
+		Scheme: splits[0],
+		Auth:   splits[1],
+	}, nil
+
+	// auth_info is in config and structured correctly - but does it make any sense?
+	// Either 'owner' or 'schema' was set but not both - this seems like a failed attempt
+	// (e.g. ':MyUser' which omit the schema, or ':' omitting both)
+	//if owner == "" || schema == "" {
+	//	return nil, fmt.Errorf("auth_info expected format is 'schema:auth'")
+	//}
+}
+
+// SchemeID composes a scheme and id
+type SchemeID struct {
+	Scheme string
+	ID     string
+}
+
+// ParseSchemeID parses input such as "scheme:id"
+func ParseSchemeID(schemeID string) (*SchemeID, error) {
+	splits := strings.Split(schemeID, ":")
+	if len(splits) != 2 {
+		return nil, fmt.Errorf("schemeOwner expected format is 'schema:owner'")
+	}
+
+	return &SchemeID{
+		Scheme: splits[0],
+		ID:     splits[1],
+	}, nil
+
+	// znode_owner is in config and structured correctly - but does it make any sense?
+	// Either 'owner' or 'schema' was set but not both - this seems like a failed attempt
+	// (e.g. ':MyUser' which omit the schema, or ':' omitting both)
+	//if owner == "" || schema == "" {
+	//	return nil, fmt.Errorf("znode_owner expected format is 'schema:auth'")
+	//}
+}
+
+// Client exports the main API that users of this package will use
+type Client struct {
+	conn   *zk.Conn
+	acl    []ACL
+	config Config
+}
+
+// New creates a new zookeeper Client.
+func New(config Config) (*Client, error) {
+	// ensure path is suffixed and prefixed (zk requires prefix /)
+	if !strings.HasSuffix(config.BasePath, "/") {
+		config.BasePath += "/"
+	}
+	if !strings.HasPrefix(config.BasePath, "/") {
+		config.BasePath = "/" + config.BasePath
+	}
+
+	if config.DefaultID == nil {
+		config.DefaultID = &SchemeID{
+			Scheme: "world",
```

Comment:
Should "world" and "anyone" be package level constants?

Reply:
As those 2 values are tied together (there is a single possible id `world:anyone` for the world scheme), I'd rather not split them in separate constants (at least not as exported constants).

I'd really like to have helper functions for the different availables schemes, either in line with this : https://github.com/samuel/go-zookeeper/blob/master/zk/util.go#L16

Or having helper functions for creating SchemeIDs (and even predeclaring common SchemeIDs, this is what the C client API does, https://zookeeper.apache.org/doc/r3.1.2/zookeeperProgrammers.html#ZooKeeper+C+client+API), and adding a `func (id SchemeID) ToACL(perms int32) []ACL` to it.

WDYT?